src/NzbDrone.Common/Cache/CachedDictionary.cs

```diff
@@ -0,0 +1,125 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace NzbDrone.Common.Cache
+{
+
+    public class CachedDictionary<TValue> : ICachedDictionary<TValue>
+    {
+        private readonly Action _refreshFunc;
+        private readonly TimeSpan _lifeTime;
+
+        private DateTime _lastRefreshed = DateTime.MaxValue;
+        private Dictionary<string, TValue> _items = new Dictionary<string, TValue>();
+
+        public CachedDictionary(Action refreshFunc, TimeSpan lifeTime)
+        {
+            _refreshFunc = refreshFunc;
+            _lifeTime = lifeTime;
+        }
+
+        public CachedDictionary(Func<Dictionary<string, TValue>> fetchFunc, TimeSpan lifeTime)
+        {
+            _refreshFunc = () => Update(fetchFunc());
+            _lifeTime = lifeTime;
+        }
+
+        protected bool IsExpired(TimeSpan lifeTime)
+        {
+            return _lastRefreshed == DateTime.MaxValue || _lastRefreshed.Add(lifeTime) < DateTime.UtcNow;
+        }
+
+        public void RefreshIfExpired()
+        {
+            RefreshIfExpired(_lifeTime);
+        }
+
+        public void RefreshIfExpired(TimeSpan lifeTime)
+        {
+            if (IsExpired(lifeTime))
+            {
+                Refresh();
+            }
+        }
+
+        public void Refresh()
+        {
+            _refreshFunc();
+            _lastRefreshed = DateTime.UtcNow;
+        }
+
+        public void Update(Dictionary<string, TValue> items)
+        {
+            if (items != null)
+            {
+                _items = items;
```

Comment:
Provide an empty Dictionary and it will. Update(null) is intended to say 'sry couldn't fetch it', but reset the ttl so we don't spam.